### ISSUE TITLE
Synchronize/sequence item processing in a stream

### ISSUE THREAD

**Issue by petrusek** [ReactiveX/RxPY#571]
Hello,
I'm trying to sequence an asyncio worker which is receiving its work through a subject. The worker is running as a side effect in a pipe operator. By sequencing I mean that no new work should start before the previous work is done.

I've managed to do this partially when the work is added "_nicely_" with `await add(1)`. Unfortunately this will not always be the case in a bigger async project. Multiple callbacks could be called shortly after each other, and even though each of them awaits, the work would get overlapped. I simulate this with `asyncio.gather()` in the example.

**I would like to delay the emitted items until previous items are done processing.**

## Example

```python
import asyncio
from asyncio.futures import Future
import time
from typing import NamedTuple
import rx.operators as ops
from rx.subject import Subject
from rx.scheduler.eventloop import AsyncIOScheduler


start = time.time()
def ts():
    return f"{time.time() - start:.3f}"

ACTION_DURATION = 2.0
Data = NamedTuple('Data', [('id', int), ('future', Future[int])])
subject = Subject()

async def async_action(id: int = 0):
    """ Some async processing, like sending/writing data. """
    print(f"{ts()} Async action  started {id}")
    await asyncio.sleep(ACTION_DURATION)
    print(f"{ts()} Async action finished {id}")
    return id


async def setup():
    """ Setting up the Rx subject to make sure all work is done
        in sequence without overlaps.
    """

    def sequencer(data: Data):
        task = asyncio.create_task(async_action(data.id))
        task.add_done_callback(
            lambda f: data.future.set_result(f.result())
        )
        return task

    subject.pipe(
        ops.flat_map(sequencer)
    ).subscribe(
        scheduler=AsyncIOScheduler(asyncio.get_event_loop())
    )


async def add(id: int):
    """ Add "some work" and await until it is done. """
    future = asyncio.Future()
    subject.on_next(Data(id, future))
    return await future


async def main():
    await setup()

    # add work in any fashion (code I cannot influence)
    print(f"{ts()} before 1")
    await add(1)
    print(f"{ts()} after  1")

    await asyncio.sleep(0.1)

    print(f"{ts()} before 2")
    await add(2)
    print(f"{ts()} after  2")

    print(f"{ts()} before 4, 5, 6")
    await asyncio.gather(add(4), add(5), add(6))
    print(f"{ts()} after  4, 5, 6")


asyncio.run(main())
```

## Output
```
$ python3 rxsync.py
0.007 before 1
0.010 Async action  started 1
2.014 Async action finished 1
2.014 after  1
2.114 before 2
2.115 Async action  started 2
4.117 Async action finished 2
4.117 after  2
4.117 before 4, 5, 6
4.119 Async action  started 4
4.119 Async action  started 5
4.119 Async action  started 6
6.120 Async action finished 4
6.120 Async action finished 5
6.120 Async action finished 6
6.121 after  4, 5, 6
```

The problem is 4, 5, 6.
What I would like, is to postpone 5 until 4 is done, and to postpone 6 until 5 is done... so there are no overlaps.

Thank you for any help.

_Note: In RxJs I've managed to solve this with the `concatMap` operator and awaiting inside its function. But in RxPY there is no concat_map and the only operator I've found that accepts a Future is flat_map. I went through the docs and also issues here on GitHub, but everything I've found and tried didn't seem to solve my issue. I saw a note that `concat_map() == map().concat()` in #129, tried using `rx.from_future` but still couldn't figure this out. Maybe it is a feature request, not sure :-)._

**Comment by MainRo** [comment]
Do you have a way to control the emission of the source items (Probably no, based on your comment "code I cannot influence")? If it is possible then some kind of backpressure handling must be implemented between the rxpy part and the source.

If this is not possible, then you need to queue all incoming items, and unqueue them when each previous item is processed.

On your example the advantage of rxpy is not obvious. If there are not additional rxpy operators being used, one option is to keep all this part as coroutines and use rxpy only after the side effects. BTW as you saw, the support of async functions in operators is very limited in rxpy. So a solution will probably keep the asyncio operations outside of the rxpy pipeline.

**Comment by petrusek** [comment]
Hi, thank you for your answer.

Well, properly queueing, un-queueing, processing, and handling all cases etc. can be programmed (I did it a few times already in js/typescript - in projects where I couldn't use RxJS), but when I was able to replace all that code with just a very clean use of the `subject` pattern and one `flatMap` operator (in JS), I was like "wow" and started loving Rx 🥰

> On your example the advantage of rxpy is not obvious.

True... it doesn't seem as much, but sequencing with just one operator would save quite a bit of code. It also has the ability to add more operators later (e.g. throttling), and since we already use Rx in the project, it is just a nice fit.

> ... the support of async functions in operators is very limited in rxpy

Regarding this, I would like to ask (since I'm relatively new to async Python3) if there is some technical reason why, or it is "just not yet". E.g. if I would like to try, is it possible to program such an operator in Python (like `concatMap` and `mergeMap` as in RxJS)? Is it worth trying?
I've seen `aioreactive` which is completely async but it also lacks these operators.
